src/main/java/bio/terra/service/dataset/DatasetDao.java

```diff
@@ -47,29 +54,143 @@ public DatasetDao(DataRepoJdbcConfiguration jdbcConfiguration,
-            throw new InvalidDatasetException("Cannot create dataset: " + dataset.getName(), e);
+    /**
+     * Lock the dataset object before doing something with it (e.g. create, delete, bulk file load).
+     * This method returns successfully when there is a dataset object locked by this flight, and throws an exception
+     * in all other cases. Below is an outline of the logic flow of this method.
+     *     1. Check if the dataset is already locked.
+     *         a. If it's locked by THIS flight, then this must be a flight recovery, so return without error.
+     *         b. If it's locked by ANOTHER flight, then throw an exception.
+     *     2. Depending on the LockBehaviorFlag specified, throw an exception if the dataset does/not already exist.
+     *     3. Either create a new dataset or update the existing one, and give this flight the lock.
+     * @param datasetName name of the dataset to lock, this is a unique column
+     * @param defaultProfileId default profile id of the dataset, this is a required column for creating a new dataset
+     * @param flightId flight id that wants to lock the dataset
+     * @param lockFlag flag specifying the lock behavior (i.e. whether to throw an exception if dataset does/not exist)
+     * @throws DatasetLockException if the dataset is locked by another flight,
+     *                              or if it does/not exist according to the LockBehaviorFlag specified
+     */
+    @Transactional(propagation =  Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public void lock(String datasetName, UUID defaultProfileId, String flightId, LockBehaviorFlags lockFlag) {
+        if (flightId == null) {
+            throw new DatasetLockException("Locking flight id cannot be null");
+        }
+
+        // lookup the dataset
+        String sql = "SELECT flightid FROM dataset WHERE name = :name";
+        MapSqlParameterSource params = new MapSqlParameterSource().addValue("name", datasetName);
+        List<String> selectResult = jdbcTemplate.query(sql, params,
+            (rs, rowNum) -> rs.getString(1));
+
+        boolean datasetExists = (selectResult.size() == 1);
+        String foundFlightId = datasetExists ? selectResult.get(0) : null;
+        boolean datasetLocked = foundFlightId != null;
+
+        // if the dataset is already locked...
+        if (datasetLocked && flightId.equals(foundFlightId)) {
+            // ...by THIS flight, then this must be a flight recovery
+            logger.debug("dataset already locked by THIS flight, must be a flight recovery");
+            return; // return without error
+        } else if (datasetLocked && !flightId.equals(foundFlightId)) {
+            // ...by ANOTHER flight, then throw an exception
+            logger.debug("dataset already locked by ANOTHER flight, error");
+            throw new DatasetLockException("Dataset is locked by another flight: " + foundFlightId);
+        }
```

Comment:
I think this would be easier to follow like this:
```suggestion
if (datasetLocked) {
    if (flightId.equals(foundFlightId)) {
            // ...by THIS flight, then this must be a flight recovery
            logger.debug("dataset already locked by THIS flight, must be a flight recovery");
            return; // return without error
    }
    // ...by ANOTHER flight, then throw an exception
    logger.debug("dataset already locked by ANOTHER flight, error");
    throw new DatasetLockException("Dataset is locked by another flight: " + foundFlightId);
}
```

Reply:
I don't think I need this chunk of logic anymore now that I've separated out the create case.